Add Installation and Configuration for Benchmark section (#4124)
opensearch-project/documentation-website: _benchmark/configuring-benchmark.md

| `async.debug` | Boolean | Enables debug mode on OpenSearch Benchmark's asyncio event loop. Default is `false`. |
| `passenv` | String | A comma-separated list of environment variable names that should be passed to OpenSearch for processing. |

## node

This section contains node-specific information that can be customized according to the needs of your cluster.

| Parameter | Type | Description |
| :---- | :---- | :---- |
| `root.dir` | String | The directory that stores all OpenSearch Benchmark data. OpenSearch Benchmark assumes control over this directory and all its subdirectories. |
| `src.root.dir` | String | The directory from which the OpenSearch source code and any OpenSearch plugins are called. Only relevant for benchmarks from [sources](#source). |

## source

This section contains more details about the OpenSearch source tree.

| Parameter | Type | Description |
| :---- | :---- | :---- |
| `remote.repo.url` | URL | The URL from which to check out OpenSearch.  Default is `https://github.com/opensearch-project/OpenSearch.git`.
| `opensearch.src.subdir` | String | The local path relative to `src.root.dir` of the OpenSearch search tree. Default is `OpenSearch`. 

> | `cache` | Boolean | Enables OpenSearch's internal source artifact cache `opensearch*.tar.gz` and any plugin zip files. Artifacts are cached based on their git revision. Default is `true`. |

Author: ```suggestion
| `cache` | Boolean | Enables OpenSearch's internal source artifact cache, `opensearch*.tar.gz`, and any plugin zip files. Artifacts are cached based on their Git revision. Default is `true`. |
```

| :---- | :---- | :---- |
| `local.dataset.cache` | String | The directory in which benchmark data sets are stored. Depending on the benchmarks that are run, this directory may contain hundreds of GB of data. Default path is `$HOME/.benchmark/benchmarks/data`. |

## results_publishing

This section defines how benchmark metrics are stored.

| Parameter | Type | Description |
| :---- | :---- | :---- |
| `datastore.type` | String | If set to `in-memory` all metrics are kept in memory while running the benchmark. If set to `opensearch` all metrics are instead written to a persistent metrics store and the data
is available for further analysis. Default is `in-memory`. |
| `sample.queue.size` | Function | The number of metrics samples that can be stored in OpenSearch Benchmark’s in-memory queue. Default is `2^20`. | 
| metrics.request.downsample.factor | Integer| (default: 1): Determines how many service time and latency samples are saved in the metrics store. By default, all values are saved. If you want to, for example. keep only every 100th sample, specify `100`. This is useful to avoid overwhelming the metrics store in benchmarks with many clients. Default is `1`. |

> | `output.processingtime` | Boolean | If set to `true`, OpenSearch shows the additional metric processing time in the command line report. Default is `false`. |

Author: Need to keep metric in here for technical accuracy.

---
layout: default
title: Configuring OpenSearch Benchmark
nav_order: 7
has_children: false
---

# Configuring OpenSearch Benchmark

OpenSearch Benchmark configuration data is stored in `~/.benchmark/benchmark.ini` which is automatically created the first time OpenSearch Benchmark runs. 

The file is separated into the following sections that you can customize based on the needs of your cluster.

## meta

This section contains meta information about the configuration file.

| Parameter | Type | Description |
| :---- | :---- | :---- |
| `config.version` | Integer |  The version of the configuration file format. This property is managed by OpenSearch Benchmark and should not be changed. |

## system

This section contains global information for the current benchmark environment. This information should be identical on all machines where OpenSearch Benchmark is installed.

| Parameter | Type | Description |
| :---- | :---- | :---- |

> | `env.name` | String | The name of the benchmark environment used a meta-data in metrics documents when an OpenSearch metrics store is configured.  Only alphanumeric characters are allowed. Default is `local`. |

Reviewer: "used a meta-data" --> "used as meta-data"

> | `env.name` | String | The name of the benchmark environment used as meta-data in metrics documents when an OpenSearch metrics store is configured.  Only alphanumeric characters are allowed. Default is `local`. |

Reviewer 2: ```suggestion
| `env.name` | String | The name of the benchmark environment used as metadata in metrics documents when an OpenSearch metrics store is configured.  Only alphanumeric characters are allowed. Default is `local`. |
```

| `available.cores` | Integer | Determines the number of available CPU cores. OpenSearch Benchmark aims to create one asyncio event loop per core and distributes to clients evenly across event loops. Defaults to the number of logical CPU cores for your cluster. |
| `async.debug` | Boolean | Enables debug mode on OpenSearch Benchmark's asyncio event loop. Default is `false`. |

> | `passenv` | String | A comma-separated list of environment variable names that should be passed to OpneSearch for processing. |

Reviewer: "OpneSearch" --> "OpenSearch"

Author: Fixed

## node

> This section contains node-specific information that can be customized according to the your cluster.

Reviewer: "the your cluster." --> "your cluster."

| Parameter | Type | Description |
| :---- | :---- | :---- |

> | `root.dir` | String | The directory that stores all OpenSearch Benchmark data. OpenSearch Benchmark assumes control over this directory and all subdirectories therein. |

Reviewer 2: ```suggestion
| `root.dir` | String | The directory that stores all OpenSearch Benchmark data. OpenSearch Benchmark assumes control over this directory and all its subdirectories. |
```

> | `src.root.dir` | String | The directory where the OpenSearch source code and any OpenSearch plugins are called. Only relevant for benchmarks from [sources](#source). |

Reviewer 2: ```suggestion
| `src.root.dir` | String | The directory from which the OpenSearch source code and any OpenSearch plugins are called. Only relevant for benchmarks from [sources](#source). |
```

## source

This section contains more details about the OpenSearch source tree.

| Parameter | Type | Description |
| :---- | :---- | :---- |

> | `remote.repo.url` | URL | The URL from which to checkout OpenSearch.  Default is `https://github.com/opensearch-project/OpenSearch.git`.

Reviewer 2: ```suggestion
| `remote.repo.url` | URL | The URL from which to check out OpenSearch.  Default is `https://github.com/opensearch-project/OpenSearch.git`.
```

| `opensearch.src.subdir` | String | The local path relative to `src.root.dir` of the OpenSearch search tree. Default is `OpenSearch`. 
| `cache` | Boolean | Enables OpenSearch's internal source artifact cache `opensearch*.tar.gz` and any plugin zip files. Artifacts are cached based on their git revision. Default is `true`. |
| `cache.days` | Integer | The number of days for which an artifact should be kept in the source artifact cache. Default is `7`. |

## benchmarks

This section contains the settings that can be customized in the OpenSearch Benchmark data directory.

| Parameter | Type | Description |
| :---- | :---- | :---- |

> | `local.dataset.cache` | String | The directory in which benchmark data sets are stored. Depending on the benchmarks that are run, this directory may contain hundreds of GB of data. |

Reviewer: Maybe we could provide that the default is `$HOME/.benchmark/benchmarks/data`

## results_publishing

This section defines how benchmark metrics are stored.

| Parameter | Type | Description |
| :---- | :---- | :---- |

> | `datastore.type` | String | If set to `in-memory` all metrics will be kept in memory while running the benchmark. If set to `opensearch` all metrics will instead be written to a persistent metrics store and the data

Reviewer 2: ```suggestion
| `datastore.type` | String | If set to `in-memory` all metrics are kept in memory while running the benchmark. If set to `opensearch` all metrics are instead written to a persistent metrics store and the data
```

> are available for further analysis. Default is `in-memory`. |

Reviewer 2: ```suggestion
is available for further analysis. Default is `in-memory`. |
```

| `sample.queue.size` | Function | The number of metrics samples that can be stored in OpenSearch Benchmark’s in-memory queue. Default is `2^20`. | 

> | metrics.request.downsample.factor | Integer| (default: 1): Determines how many service time and latency samples should be kept in the metrics store. By default all values will be kept. If you want to, for example. keep only every 100th sample, specify `100`. This is useful to avoid overwhelming the metrics store in benchmarks with many clients. Default is `1`. |

Reviewer 2: ```suggestion
| metrics.request.downsample.factor | Integer| (default: 1): Determines how many service time and latency samples are saved in the metrics store. By default, all values are saved. If you want to, for example. keep only every 100th sample, specify `100`. This is useful to avoid overwhelming the metrics store in benchmarks with many clients. Default is `1`. |
```

> | `output.processingtime` | Boolean | If set to `true`, OpenSearch will show the additional metric processing time in the command line report. Default is `false`. |

Reviewer 2: ```suggestion
| `output.processingtime` | Boolean | If set to `true`, OpenSearch shows the additional metric processing time in the command line report. Default is `false`. |
```

### `datastore.type` parameters

When `datastore.type` is set to `opensearch`, the following reporting settings can be customized.

| Parameter | Type | Description |
| :---- | :---- | :---- |

> | `datastore.host` | IP address | The hostname of the metrics store, for example `124.340.200.22`. |

Reviewer 2: ```suggestion
| `datastore.host` | IP address | The hostname of the metrics store, for example, `124.340.200.22`. |
```

| datastore.port| Port | The port number of the metrics store, for example, `9200`. |
| `datastore.secure` | Boolean | If set to `false`, OpenSearch assumes a HTTP connection. If set to true, it assumes a HTTPS connection. |
| `datastore.ssl.verification_mode` | String | When set to the default `full`, the metric store’s SSL certificate is checked. To disable certificate verification set this value to `none`. |
| `datastore.ssl.certificate_authorities` | String | Determines the path on the local file system to the certificate authority’s signing certificate.
| `datastore.user` | Username |  Sets the name of the OpenSearch user for the metrics store. |

> | `datastore.password`: | String | Sets the password of the OpenSearch user for the metrics store. Alternatively, this password can be configured using the `OPENSEARCH_REPORTING_DATASTORE_PASSWORD` environment variable, which avoids storing credentials in a plain text file. The environment variable will take precedence over the config file if both define a password. |

Reviewer: We can swap `OPENSEARCH_REPORTING_DATASTORE_PASSWORD` for `OSB_RESULTS_PUBLISHING_PASSWORD` to stay in line with "Results_Publishing" section.

| `datastore.number_of_replicas` | Integer |  The number of replicas each primary shard has. Any updates to this setting after initial index creation will only be applied to any new `opensearch-* `indexes. Default is the [OpenSearch static index value]({{site.url}}{{site.baseurl}}/api-reference/index-apis/create-index/#static-index-settings). |

### Examples

You can use the following examples to set reporting values in your own cluster.

This example defines an unprotected metrics store in the local network:

```
[results_publishing]
datastore.type = opensearch
datastore.host = 192.168.10.17
datastore.port = 9200
datastore.secure = false
datastore.user =
datastore.password =
```

This example defines a secure connection to a metrics store in the local network with a self-signed certificate:

```
[results_publishing]
datastore.type = opensearch
datastore.host = 192.168.10.22
datastore.port = 9200
datastore.secure = true
datastore.ssl.verification_mode = none
datastore.user = user-name
datastore.password = the-password-to-your-cluster
```

## workloads

> This section defines how workloads are retrieved. All keys are read by OpenSearch using the syntax `<<workload-repository-name>>.url`, which you can select using the OpenSearch Benchmark CLI  `--track-repository=workload-repository-name"` option. By default, OpenSearch chooses the track repository using the `default.url`, `https://github.com/opensearch-project/opensearch-benchmark-workloads`.

Reviewer: 1. `--track-repository` --> `--workload-repository`
2. `OpenSearch chooses the track repository` --> `workload repository`

## teams

> What is Benchmark teams equivalent?

Author: Removed

| `release.cache` | Boolean | Determines whether newly released OpenSearch versions should be cached locally. |

## Proxy configurations

OpenSearch automatically downloads all the necessary proxy data for you, including:

- OpenSearch distributions when you specify `--distribution-version=<OPENSEARCH-VERSION>`. 
- OpenSearch source code when you specify a Git revision number, for example `--revision=1e04b2w`. 
- Any metadata tracked from the [OpenSearch GitHub repository](https://github.com/opensearch-project/OpenSearch).

As of OpenSearch Benchmark 0.5.0, only `http_proxy` is supported.
{: .warning}

You can use an `http_proxy` to connect OpenSearch Benchmark to a specific proxy and connect the proxy to a Benchmark workload. To add the proxy: 

1. Add your proxy URL to your shell profile:

   ```
   export http_proxy=http://proxy.proxy.org:4444/
   ```

2. Source your shell profile and verify that the proxy URL is correctly set:

   ```
   source ~/.bash_profile ; echo $http_proxy
   ```

3. Configure Git to connect to your proxy. For more information, see the [Git Documentation](https://git-scm.com/docs/git-config).

   ```
   git config --global http_proxy $http_proxy
   ```

> 4. Use `git clone` to clone the workloads repository. If the proxy configured correctly, the clone will be successful.

Reviewer 2: ```suggestion
4. Use `git clone` to clone the workloads repository. If the proxy configured correctly, the clone is successful.
```
